egs/librispeech/ASR/tdnn_lstm_ctc/asr_datamodule.py

```diff
@@ -473,3 +486,240 @@ def test_other_cuts(self) -> CutSet:
+            world_size=1,
+        )
+        logging.info("About to create train dataloader")
+
+        # 'seed' is derived from the current random state, which will have
+        # previously been set in the main process.
+        seed = torch.randint(0, 100000, ()).item()
+        rank, world_size = None, None
+        if torch.distributed.is_initialized():
+            rank, world_size = (
+                torch.distributed.get_rank(),
+                torch.distributed.get_world_size(),
+            )
+
+        train_dl = DataLoader(
+            IterableDatasetWrapper(dataset=train, sampler=train_sampler),
+            num_workers=self.args.num_workers,
+            batch_size=None,
+            worker_init_fn=make_worker_init_fn(
+                rank=rank, world_size=world_size, seed=seed
+            ),
+        )
+
+        return train_dl
+
+    def valid_dataloaders(self, cuts_valid: CutSet) -> DataLoader:
+        logging.info("About to create dev dataset")
+        if self.args.input_strategy == "OnTheFlyFeatures":
+            validate = K2SpeechRecognitionDataset(
+                input_strategy=OnTheFlyFeatures(Fbank(FbankConfig(num_mel_bins=80)))
+            )
+        else:
+            validate = K2SpeechRecognitionDataset()
+        valid_sampler = DynamicBucketingSampler(
+            cuts_valid,
+            max_duration=self.args.max_duration,
+            shuffle=False,
+        )
+        logging.info("About to create dev dataloader")
+        valid_dl = DataLoader(
+            validate,
+            sampler=valid_sampler,
+            batch_size=None,
+            num_workers=2,
+            persistent_workers=False,
+        )
+
+        return valid_dl
+
+    def test_dataloaders(self, cuts: CutSet) -> DataLoader:
+        logging.debug("About to create test dataset")
+        test = K2SpeechRecognitionDataset(
+            input_strategy=OnTheFlyFeatures(Fbank(FbankConfig(num_mel_bins=80)))
+            if self.args.input_strategy == "OnTheFlyFeatures"
+            else eval(self.args.input_strategy)(),
+        )
+        sampler = DynamicBucketingSampler(
+            cuts,
+            max_duration=self.args.max_duration,
+            shuffle=False,
+        )
+        logging.debug("About to create test dataloader")
+        test_dl = DataLoader(
+            test,
+            batch_size=None,
+            sampler=sampler,
+            num_workers=self.args.num_workers,
+        )
+        return test_dl
+
+    @lru_cache()
+    def train_all_shuf_cuts(self) -> CutSet:
+        logging.info(
+            "About to get the shuffled and sharded cuts for full (960h) LibriSpeech using Lhotse Shar"
+        )
+        # Below we'll figure out which files to read.
+        # Since we only use either (precomputed) features or recordings,
+        # we shouldn't iterate over both at the same time.
+        shar_dir = Path(self.args.shar_dir)
+        fields = {"cuts": sorted(shar_dir.glob("cuts.*.jsonl*"))}
```

Comment:
I am a little confused about this glob stuff, is it just grabbing everything in the shar directory, and how does it work if we just want train-clean-100 or the test set?

Reply:
Yeah it only supports full LibriSpeech. But you can shar-export every piece separately and then here you’d concatenate the result of globs in each exported directory.

Reply:
OK.  IMO it might be a better practice to give the shar-dir a more specific name, like shar_train_960 or something, to indicate it is a specific dataset.